### PR TITLE
Update Review app to import `govuk-frontend` via local package

### PR DESCRIPTION
**Pull request by colinrotherham**
This PR makes the Review app a user of `govuk-frontend` as part of:

* https://github.com/alphagov/govuk-frontend/issues/3280

For the performance work all source code and build artifacts can now be resolved from a single package

Additionally, the Review app now follows our own documentation:
https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/

These areas in particular:

## Import CSS, assets and JavaScript
1. [Simplify Sass import paths](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#simplify-sass-import-paths)
2. [Import JavaScript using a bundler](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#import-javascript-using-a-bundler)

Which means the Review app now imports `govuk-frontend` ESM JavaScript for the first time, without having to navigate to `../../../src` sources or rely on a separate review-specific UMD bundle

## Package directory changes
Previously `npm link` only resolved the pre-built packages files but they're now consolidated:

### Before

```console
├── app
│   ├── src
│   └── dist
└── package
    ├── govuk
    ├── govuk-esm
    ├── govuk-prototype-kit
    ├── package.json
    └── README.md
```

### After

```console
└── packages
    ├── govuk-frontend
    │   ├── src
    │   │   ├── govuk
    │   │   └── govuk-prototype-kit
    │   ├── dist
    │   │   ├── govuk
    │   │   ├── govuk-esm
    │   │   └── govuk-prototype-kit
    │   ├── package.json
    │   └── README.md
    └── govuk-frontend-review
        ├── src
        └── dist
```

~`src`~ → `packages/govuk-frontend/src`
~`package`~ → `packages/govuk-frontendpackage/dist`

To maintain backwards compatibility these directory changes aren't found in the [published `package.json`](https://github.com/alphagov/govuk-frontend/blob/review-app-user/package/dist/package.json)

### The problem with `dist`
Unfortunately the standard convention of building `src` to `dist` doesn't fit

We already reserve `./dist` for our GitHub tagged releases, do we need to move it?
Will it be confusing to have a committed `./dist` alongside `./package/src` → `./package/dist` build output?

Some options:

1. `dist` — Leave it where it is
2. ~`dist`~ → `package/release` — Maintain two build directories under `./package`
3. ~`dist`~ → `package/dist/release` — Consolidate single build directory under `./package`

This PR has currently picked 1) from the list above

### CODE REVIEW

File: .github/workflows/diff.yaml
Comment (domoscargin) [inline .github/workflows/diff.yaml:1]: I think I'm missing some context for the rationale behind renaming `dist` to `release` (and `package` to `dist`). I'm not immediately sure of the difference between distribution and release just from naming (I've seen differing uses of those words, but maybe opinion is solidifying around how we use 'em?).

Not against it at all (it links up to our "release" archives at the very least), just think we need some explanation of it in the PR's description/commit messages.

File: .github/workflows/tests.yml
Comment (domoscargin) [inline .github/workflows/tests.yml:167]: Here's an example of the slight confusion in naming - the description here referring to "distribution", the  task referring to "release".

File: .prettierignore
Comment (domoscargin) [inline .prettierignore:13]: Interesting, do we have a base directory level `release` folder anywhere at the mo?

File: config/paths.js
Comment (domoscargin) [inline config/paths.js:16]: Again, slightly confusing wording here, since there's a `package/dist` folder as well.

File: docs/contributing/application-architecture.md
Comment (domoscargin) [inline docs/contributing/application-architecture.md:15]: First added 20 June 2017.

🖖

File: docs/contributing/application-architecture.md
Comment (domoscargin) [inline docs/contributing/application-architecture.md:32]: First commited 19 May, 2017.

F

File: app/package.json
Comment (domoscargin) [inline app/package.json:27]: Neat.

Something to remember if we ever have to account for v4 and v5 in the review app, but it should be pretty blatant.

File: app/src/stylesheets/app-ie8.scss
Comment (domoscargin) [inline app/src/stylesheets/app-ie8.scss:1]: Nice, can't stand a bunch of directory faff, this is much neater.

**Review by domoscargin — COMMENTED** [review]
Awesome stuff.

Instinctively, this feels like quite a big change just to optimise the performance monitoring work. Is there an intermediate step we can take, or is this pretty much all or nothing?

I'm not strongly opposed at all, though I am slightly iffy on the reorganisation/naming - I feel like most devs would immediately be familiar with a base level `src` folder and tucking it into `package` might obscure that a bit. Similar vibes with `dist`. And the name "package" rather than "dist" just feels like a natural fit for npm publishing, 'cuz that's all packages.

Basically think it just boils down to the PR needing a bit more of the rationale in the description maybe?

File: .github/workflows/tests.yml
Comment (colinrotherham) [reply .github/workflows/tests.yml:167]: Yeah this needs some work doesn't it

Ideally we have a single consolidated distribution via `npm publish` rather than another separate one for GitHub. Seems a shame that `npm install govuk-frontend` doesn't include those fab precompiled files

I didn't want to go that far in this PR though so I'm open to better naming/directory conventions

File: .prettierignore
Comment (colinrotherham) [reply .prettierignore:13]: Nah we don't but this matches `/package/release/` too

File: app/package.json
Comment (colinrotherham) [reply app/package.json:27]: Thanks @domoscargin

This is the reason for the directory moves 😣 

The top-level project is named [govuk-frontend-repository](https://github.com/alphagov/govuk-frontend/blob/main/package.json#L3)
But the nested package is named [govuk-frontend](https://github.com/alphagov/govuk-frontend/blob/main/package/package.json#L2)

Hence the move to ~`./src`~ → `./package/src`

Otherwise inside `node_modules` we can only see package-relative built files (not the beautiful source)

Built package only

Comment (colinrotherham) [reply app/package.json:27]: > Something to remember if we ever have to account for v4 and v5 in the review app, but it should be pretty blatant.

Yeah we can still install from the registry using:

```
npm install govuk-frontend@4.4.0 --save --workspace app
```

Confirm with `npm ls govuk-frontend` to see the registry package alongside the local package:

```console
├─┬ govuk-frontend-config@ -> ./shared/config
│ └── govuk-frontend@4.5.0 deduped -> ./package
├─┬ govuk-frontend-review@ -> ./app
│ └── govuk-frontend@4.4.0
├─┬ govuk-frontend-webpack-boilerplate@ -> ./docs/examples/webpack
│ └── govuk-frontend@4.5.0 deduped -> ./package
└── govuk-frontend@4.5.0 -> ./package
```

Clever eh?

Bit of a pain involved (this PR) but it sets us up for the future

**Comment by romaricpascal** [comment]
Having the app consume `govuk-frontend` itself, however, I'm with Brett with it looking like a big change to enable the performance work. I'm not clear about what it enables that was blocked before, so I'd welcome some more details about that, please 😄 

I'm wondering if there's a shorter route we may take to enable us to ship v5 (even if it means more manual checks, less isolation between the packages while we ship it), before we look into finishing re-organising our internal repository structure once it's shipped 😊

**Comment by colinrotherham** [comment]
@romaricpascal Sounds good, give me a nudge when you're free

File: config/paths.js
Comment (colinrotherham) [reply config/paths.js]: I've removed the word "distribution" to help clarify things

File: .github/workflows/diff.yaml
Comment (colinrotherham) [reply .github/workflows/diff.yaml]: Caught up with @romaricpascal and @36degrees today

Since the useful work is moving `govuk-frontend` we can leave `./dist` at the project root for now

File: .github/workflows/tests.yml
Comment (colinrotherham) [reply .github/workflows/tests.yml]: We're going to [keep `./dist` at project root](https://github.com/alphagov/govuk-frontend/pull/3491#discussion_r1173953748) for now

File: .prettierignore
Comment (colinrotherham) [reply .prettierignore]: We're going to [keep `./dist` at project root](https://github.com/alphagov/govuk-frontend/pull/3491#discussion_r1173953748) for now

File: package.json
Comment (domoscargin) [reply package.json:23]: I don't mind the `preview, package` naming here, though it's a bit of a shame to add _another_ name to argue over. Here's my 2p: `--names \"app,pkg\"`.

**Review by domoscargin — COMMENTED** [review]
I think this mostly looks hunky dory, thanks for walking me through it.

I've added a few comments here and there.

File: app/tasks/assets.mjs
Comment (colinrotherham) [reply app/tasks/assets.mjs:9]: Yeah good shout

What if we drop the unique names commit and not worry about it?

We've got the colour-coded prefixes from [`concurrently`](https://www.npmjs.com/package/concurrently) instead maybe

Comment (colinrotherham) [reply app/tasks/assets.mjs:9]: Yeah maybe worrying about task names can wait for another day?

Like you've found, we've now got to worry about npm scripts too

AND the generic **gulpfile.mjs** ones: `dev`, `watch`, `scripts`, `styles` 🤦‍♂️

Comment (domoscargin) [reply app/tasks/assets.mjs:9]: That works for me - the colour-coded prefixes were more helpful to me at a glance anyway, and the prefixes themselves give enough differentiation for folks who can't see the colour for whatever reason.

File: package.json
Comment (domoscargin) [reply package.json:23]: The last tool like this I remember using is `npm-run-all`, which I've just seen seen was last updated in 2018, so I don't think I can offer you much better 🤣 

Looks like the maintainer is interested in the fixed-width name columns, so hopefully a feature down the line.

My comments above aren't blocking, for sure.

File: .gitignore
Comment (colinrotherham) [reply .gitignore]: All sorted, we now have the `packages` directory structure in this PR 🙌 

But because we can still run `npm publish` from `packages/govuk-frontend/dist` (with its own **package.json**) it keeps everything backwards compatible until we're ready for:

* https://github.com/alphagov/govuk-frontend/pull/3498

File: app/tasks/assets.mjs
Comment (colinrotherham) [reply app/tasks/assets.mjs]: I've rolled back all the task renames

File: package.json
Comment (colinrotherham) [reply package.json]: This comment is marked outdated but only because it's moved to `packages/govuk-frontend-review`

I've gone with your `--names \"app,pkg\"` suggestion ✅

File: app/tasks/assets.mjs
Comment (domoscargin) [reply app/tasks/assets.mjs]: Yeah, think this looks good to me. Can always review down the line.

File: .gitignore
Comment (domoscargin) [reply .gitignore]: This looks grand!

Should we do a release/pre-release dry run before merging to make sure the branches/files generate OK?

Comment (colinrotherham) [reply .gitignore]: Yeah we should

Since we're referring to `govuk-frontend` in lots of places we'll want to merge this one first:

* https://github.com/alphagov/govuk-frontend/pull/3558

Otherwise any support/prerelease branches (or merged release PRs) will have **package-lock.json** referencing the previous `"version"` string. First person to run `npm install` will see npm fix them all

**Review by romaricpascal — COMMENTED** [review]
That looks so tidy now 😍 Thanks for your patience with this piece of work and going through it with us to explain why it needed to happen now 😊 

Left a couple of comments, but nothing seems major 🥳 On top of them 3 things come to mind before the merge:
- I'm wondering if the "Set up separate scripts, styles and watch" commit is in the right PR. It seems more linked to #3498 
- we should add a way, if only for a while, to direct people to the new places, as they're used to looking at the sources in `src` and the app in `app`. We mentioned README.md files in the old locations, which sound an easy way to sort this out, or symlinks. Does any have particular drawbacks (maybe in terms of rebasing)? I think these pointers should be part of this PR so all the info is available once merged.

We'll also need to rebase the outstanding PRs once merged, hopefully Git will be fine with all that movement 😊

File: docs/examples/webpack/webpack.config.js
Comment (colinrotherham) [reply docs/examples/webpack/webpack.config.js:7]: Totally forgot I already added `package.json` as a package export:

```suggestion
const frontendPath = dirname(require.resolve('govuk-frontend/package.json'))
```

Comment (romaricpascal) [reply docs/examples/webpack/webpack.config.js:7]: That's a neat way to fix it! 🙌🏻

File: package.json
Comment (colinrotherham) [reply package.json]: Done ✅

**Comment by colinrotherham** [comment]
Hi @romaricpascal @domoscargin

This PR was split out into related-but-separate PRs for:

* https://github.com/alphagov/govuk-frontend/pull/3560
* https://github.com/alphagov/govuk-frontend/pull/3562
* https://github.com/alphagov/govuk-frontend/pull/3491 ← current PR
* https://github.com/alphagov/govuk-frontend/pull/3498

Do you see anything missing?

I've added a README (and symlinks) to explain the `src` move too:
https://github.com/alphagov/govuk-frontend/tree/review-app-user/src

File: src/README.md
Comment (domoscargin) [inline src/README.md:3]:
```suggestion
GOV.UK Frontend source code has moved to the [`govuk-frontend` package directory](../packages/govuk-frontend).
```
Extreme nitpick: `packages` makes it sound like there's a `packages` subfolder somewhere. Singular package makes it sound as though it's the directory of the `govuk-frontend` package.

This is extremely pedantic... :rofl:

**Review by domoscargin — APPROVED** [review]
I think this looks great - thanks for breaking up the PRs and keeping everything in sync, @colinrotherham, it's been really helpful.

Would appreciate another pair of eyes on this before merging, but from our discussion it didn't seem like there'd be any major objections.

File: src/README.md
Comment (colinrotherham) [reply src/README.md:3]: Ha, yup, I lost a good couple of minutes toying with other options (another one below)

```suggestion
GOV.UK Frontend source code has moved to the [packages `govuk-frontend` directory](../packages/govuk-frontend).
```

I'm going to rebase with https://github.com/alphagov/govuk-frontend/pull/3561 anyway so I'll change it to yours 😊

**Comment by domoscargin** [comment]
Also worth popping a reminder here that this PR means we have to update our Decision on repo structure: https://github.com/alphagov/govuk-design-system-architecture/pull/24

(I see @36degrees has just added that to the board as well)

File: src/README.md
Comment (colinrotherham) [reply src/README.md]: Done ✅

**Comment by romaricpascal** [comment]
Agree with Brett on notifying people before we merge it, and making sure we don't wait too long before rebasing the PRs that still need rebasing as well 😊

**Comment by colinrotherham** [comment]
> Agree with Brett on notifying people before we merge it, and making sure we don't wait too long before rebasing the PRs that still need rebasing as well 😊

Thanks @romaricpascal

We rebased all these earlier:

[cssnano-autoprefixer](https://github.com/alphagov/govuk-frontend/compare/cssnano-autoprefixer) ([#3243](https://github.com/alphagov/govuk-frontend/pull/3243))
[delete-package-dist](https://github.com/alphagov/govuk-frontend/compare/delete-package-dist) ([#3498](https://github.com/alphagov/govuk-frontend/pull/3498))
[remove-compat-mode-legacy-colours](https://github.com/alphagov/govuk-frontend/compare/remove-compat-mode-legacy-colours) ([#3576](https://github.com/alphagov/govuk-frontend/pull/3576))
[remove-compat-mode-legacy-fonts](https://github.com/alphagov/govuk-frontend/compare/remove-compat-mode-legacy-fonts) ([#3574](https://github.com/alphagov/govuk-frontend/pull/3574))
[remove-ie8-css](https://github.com/alphagov/govuk-frontend/compare/remove-ie8-css) ([#3572](https://github.com/alphagov/govuk-frontend/pull/3572))
[remove-ie8-js](https://github.com/alphagov/govuk-frontend/compare/remove-ie8-js) ([#3570](https://github.com/alphagov/govuk-frontend/pull/3570))
[remove-typography-use-rem-setting](https://github.com/alphagov/govuk-frontend/compare/remove-typography-use-rem-setting) ([#3575](https://github.com/alphagov/govuk-frontend/pull/3575))
[review-app-user](https://github.com/alphagov/govuk-frontend/compare/review-app-user) ([#3491](https://github.com/alphagov/govuk-frontend/pull/3491))

**Pull request merged**
